Index: kernel/iscsi.h
===================================================================
--- 1/kernel/iscsi.h.p	2008-11-11 12:00:45.000000000 -0500
+++ 2/kernel/iscsi.h	2008-11-11 12:00:59.000000000 -0500
@@ -241,8 +241,8 @@ struct iscsi_conn {
 	u32 write_offset;
 	int write_state;
 
-	struct crypto_tfm *rx_digest_tfm;
-	struct crypto_tfm *tx_digest_tfm;
+	struct hash_desc rx_hash;
+	struct hash_desc tx_hash;
 	struct scatterlist hash_sg[ISCSI_CONN_IOV_MAX];
 };
 
Index: kernel/digest.c
===================================================================
--- 1/kernel/digest.c.p	2008-11-11 12:00:53.000000000 -0500
+++ 2/kernel/digest.c	2008-11-11 12:01:38.000000000 -0500
@@ -12,7 +12,8 @@
 
 void digest_alg_available(unsigned int *val)
 {
-	if (*val & DIGEST_CRC32C && !crypto_alg_available("crc32c", 0)) {
+	if (*val & DIGEST_CRC32C &&
+	    !crypto_has_alg("crc32c", 0, CRYPTO_ALG_ASYNC)) {
 		printk("CRC32C digest algorithm not available in kernel\n");
 		*val |= ~DIGEST_CRC32C;
 	}
@@ -36,15 +37,22 @@ int digest_init(struct iscsi_conn *conn)
 	if (!(conn->ddigest_type & DIGEST_ALL))
 		conn->ddigest_type = DIGEST_NONE;
 
-	if (conn->hdigest_type & DIGEST_CRC32C || conn->ddigest_type & DIGEST_CRC32C) {
-		conn->rx_digest_tfm = crypto_alloc_tfm("crc32c", 0);
-		if (!conn->rx_digest_tfm) {
+	if (conn->hdigest_type & DIGEST_CRC32C ||
+	    conn->ddigest_type & DIGEST_CRC32C) {
+		conn->rx_hash.tfm = crypto_alloc_hash("crc32c", 0,
+						      CRYPTO_ALG_ASYNC);
+		conn->rx_hash.flags = 0;
+		if (IS_ERR(conn->rx_hash.tfm)) {
+			conn->rx_hash.tfm = NULL;
 			err = -ENOMEM;
 			goto out;
 		}
 
-		conn->tx_digest_tfm = crypto_alloc_tfm("crc32c", 0);
-		if (!conn->tx_digest_tfm) {
+		conn->tx_hash.tfm = crypto_alloc_hash("crc32c", 0,
+						      CRYPTO_ALG_ASYNC);
+		conn->tx_hash.flags = 0;
+		if (IS_ERR(conn->tx_hash.tfm)) {
+			conn->tx_hash.tfm = NULL;
 			err = -ENOMEM;
 			goto out;
 		}
@@ -65,10 +73,10 @@ out:
  */
 void digest_cleanup(struct iscsi_conn *conn)
 {
-	if (conn->tx_digest_tfm)
-		crypto_free_tfm(conn->tx_digest_tfm);
-	if (conn->rx_digest_tfm)
-		crypto_free_tfm(conn->rx_digest_tfm);
+	if (conn->tx_hash.tfm)
+		crypto_free_hash(conn->tx_hash.tfm);
+	if (conn->rx_hash.tfm)
+		crypto_free_hash(conn->rx_hash.tfm);
 }
 
 /**
@@ -159,28 +167,28 @@ static inline void __dbg_simulate_data_d
 	(sg).length = (l);					\
 } while (0)
 
-static void digest_header(struct crypto_tfm *tfm, struct iscsi_pdu *pdu, u8 *crc)
+static void digest_header(struct hash_desc *hash, struct iscsi_pdu *pdu,
+			  u8 *crc)
 {
 	struct scatterlist sg[2];
-	int i = 0;
+	unsigned int nbytes = sizeof(struct iscsi_hdr);
 
-	SETSG(sg[i], &pdu->bhs, sizeof(struct iscsi_hdr));
-	i++;
+	SETSG(sg[0], &pdu->bhs, nbytes);
 	if (pdu->ahssize) {
-		SETSG(sg[i], pdu->ahs, pdu->ahssize);
-		i++;
+		SETSG(sg[1], pdu->ahs, pdu->ahssize);
+		nbytes += pdu->ahssize;
 	}
 
-	crypto_digest_init(tfm);
-	crypto_digest_update(tfm, sg, i);
-	crypto_digest_final(tfm, crc);
+	crypto_hash_init(hash);
+	crypto_hash_update(hash, sg, nbytes);
+	crypto_hash_final(hash, crc);
 }
 
 int digest_rx_header(struct iscsi_cmnd *cmnd)
 {
 	u32 crc;
 
-	digest_header(cmnd->conn->rx_digest_tfm, &cmnd->pdu, (u8 *) &crc);
+	digest_header(&cmnd->conn->rx_hash, &cmnd->pdu, (u8 *) &crc);
 	if (crc != cmnd->hdigest)
 		return -EIO;
 
@@ -189,18 +197,19 @@ int digest_rx_header(struct iscsi_cmnd *
 
 void digest_tx_header(struct iscsi_cmnd *cmnd)
 {
-	digest_header(cmnd->conn->tx_digest_tfm, &cmnd->pdu, (u8 *) &cmnd->hdigest);
+	digest_header(&cmnd->conn->tx_hash, &cmnd->pdu, (u8 *) &cmnd->hdigest);
 }
 
-static void digest_data(struct crypto_tfm *tfm, struct iscsi_cmnd *cmnd,
+static void digest_data(struct hash_desc *hash, struct iscsi_cmnd *cmnd,
 			struct tio *tio, u32 offset, u8 *crc)
 {
 	struct scatterlist *sg = cmnd->conn->hash_sg;
 	u32 size, length;
 	int i, idx, count;
+	unsigned int nbytes;
 
 	size = cmnd->pdu.datasize;
-	size = (size + 3) & ~3;
+	nbytes = size = (size + 3) & ~3;
 
 	offset += tio->offset;
 	idx = offset >> PAGE_CACHE_SHIFT;
@@ -210,7 +219,7 @@ static void digest_data(struct crypto_tf
 
 	assert(count <= ISCSI_CONN_IOV_MAX);
 
-	crypto_digest_init(tfm);
+	crypto_hash_init(hash);
 
 	for (i = 0; size; i++) {
 		if (offset + size > PAGE_CACHE_SIZE)
@@ -225,8 +234,8 @@ static void digest_data(struct crypto_tf
 		offset = 0;
 	}
 
-	crypto_digest_update(tfm, sg, count);
-	crypto_digest_final(tfm, crc);
+	crypto_hash_update(hash, sg, nbytes);
+	crypto_hash_final(hash, crc);
 }
 
 int digest_rx_data(struct iscsi_cmnd *cmnd)
@@ -252,9 +261,10 @@ int digest_rx_data(struct iscsi_cmnd *cm
 		offset = 0;
 	}
 
-	digest_data(cmnd->conn->rx_digest_tfm, cmnd, tio, offset, (u8 *) &crc);
+	digest_data(&cmnd->conn->rx_hash, cmnd, tio, offset, (u8 *) &crc);
 
-	if (!cmnd->conn->read_overflow && (cmnd_opcode(cmnd) != ISCSI_OP_PDU_REJECT)) {
+	if (!cmnd->conn->read_overflow &&
+	    (cmnd_opcode(cmnd) != ISCSI_OP_PDU_REJECT)) {
 		if (crc != cmnd->ddigest)
 			return -EIO;
 	}
@@ -268,6 +278,6 @@ void digest_tx_data(struct iscsi_cmnd *c
 	struct iscsi_data_out_hdr *req = (struct iscsi_data_out_hdr *)&cmnd->pdu.bhs;
 
 	assert(tio);
-	digest_data(cmnd->conn->tx_digest_tfm, cmnd, tio,
+	digest_data(&cmnd->conn->tx_hash, cmnd, tio,
 		    be32_to_cpu(req->buffer_offset), (u8 *) &cmnd->ddigest);
 }
